src/Spago/Config.purs

```diff
@@ -174,130 +175,212 @@ type ReadWorkspaceOptions =
-        if Array.any (_ `Path.isPrefixOf` packagePath) configPathsWithWorkspaces then
-          pure { right, left: Array.cons packagePath left }
+      misnamedConfigsList =
+        case misnamedConfigs <#> \c -> Path.quote $ (c </> "spago.yml") `Path.relativeTo` root of
+          [] -> []
+          [ one ] -> [ toDoc $ "Instead found " <> one ]
+          many -> [ toDoc "Instead found these:", indent $ toDoc many ]
+    die
+      [ toDoc $ "No " <> spagoYaml <> " found in the current directory or any of its parents."
+      , if Array.null misnamedConfigsList then
+          toDoc ""
         else
-          case readResult.config.yaml.package of
-            Nothing ->
-              pure { right, left: Array.cons packagePath left }
-            Just package -> do
-              -- Note: we migrate configs only at this point - this is because we read a whole lot of them but we are
-              -- supposed to ignore any subtrees that contain a different workspace, and those we don't want to migrate
-              doMigrateConfig configPath config
-              -- We store the path of the package, so we can treat it basically as a LocalPackage
-              pure { left, right: Array.cons (Tuple package.name { package, hasTests, path: packagePath, doc: Just config.doc }) right }
-    in
-      Array.foldM fn { right: [], left: [] } otherPackages
-
-  unless (Array.null prunedConfigs) do
-    logDebug
-      $ [ "Excluding configs that use a different workspace (directly or implicitly via parent directory's config):" ]
-      <> Array.sort (Path.quote <$> prunedConfigs)
-
-  rootPackage <- case maybePackage of
-    Nothing -> pure []
-    Just rootPackage -> do
-      rootPackage' <- rootPackageToWorkspacePackage rootPath { rootPackage, workspaceDoc }
-      pure [ Tuple rootPackage.name rootPackage' ]
-
-  let workspacePackages = Map.fromFoldable $ configsNoWorkspaces <> rootPackage
+          toDoc
+            [ Log.break
+            , indent $ toDoc $ misnamedConfigsList
+            , indent $ toDoc $ "Note that Spago config files should be named " <> spagoYaml <> ", not spago.yml."
+            , Log.break
+            , toDoc "The configuration file help can be found here https://github.com/purescript/spago#the-configuration-file"
+            ]
+      ]
+
+determineSelectedPackage
+  :: ∀ a
+   . { explicitlySelected :: Maybe PackageName
+     , inferredFromCwd :: Maybe PackageName
+     , rootPackage :: Maybe PackageName
+     , loadedPackages :: Map PackageName WorkspacePackage
+     }
+  -> Spago (Registry.RegistryEnv a) (Maybe WorkspacePackage)
+determineSelectedPackage { explicitlySelected, inferredFromCwd, rootPackage, loadedPackages } = do
+  let
+    inferredFromCwd' =
+      -- Do not auto-select the root package if Spago is being run from the root directory.
+      if inferredFromCwd == rootPackage then Nothing else inferredFromCwd
+
+    selectedName =
+      explicitlySelected <|> inferredFromCwd'
```

Comment:
As you said in the initial comment, I'm not sure we should infer the selected package from the cwd. On one hand we have always refrained from auto-selecting packages (except in very specific cases), on the other hand this looks quite sensible?

We do not auto-select if we are in the workspace folder, where it makes sense to build the whole thing. If you are in a subdirectory I think it's safe to assume that you want to refer to the package where you currently are.

All of this to say: yes sure why not